nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/QueryIoTDBRecord.java

```diff
@@ -0,0 +1,219 @@
+                    + "Chunking queries can return an unlimited number of rows. Note: Chunking is enable when result chunk size is greater than 0")
+            .defaultValue(String.valueOf(DEFAULT_IOTDB_FETCH_SIZE))
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .addValidator(StandardValidators.createLongValidator(0, FETCH_SIZE, true))
+            .required(true)
+            .build();
+
+    private static RecordSetWriterFactory recordSetWriterFactory;
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> propertyDescriptors = new ArrayList<>(super.getSupportedPropertyDescriptors());
+        propertyDescriptors.add(IOTDB_QUERY);
+        propertyDescriptors.add(IOTDB_QUERY_FETCH_SIZE);
+        propertyDescriptors.add(RECORD_WRITER_FACTORY);
+        return Collections.unmodifiableList(propertyDescriptors);
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) throws IoTDBConnectionException {
+        super.onScheduled(context);
+        // Either input connection or scheduled query is required
+        if (!context.getProperty(IOTDB_QUERY).isSet()
+                && ! context.hasIncomingConnection()) {
+            throw new ProcessException("The IoTDB Query processor requires input connection or scheduled IoTDB query");
+        }
+        recordSetWriterFactory = context.getProperty(RECORD_WRITER_FACTORY).asControllerService(RecordSetWriterFactory.class);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession processSession) throws ProcessException {
+
+        String query;
+        Charset charset;
+        FlowFile outgoingFlowFile;
+        int fetchSize = DEFAULT_IOTDB_FETCH_SIZE;
+        // If there are incoming connections, prepare query params from flow file
+        if (context.hasIncomingConnection()) {
+            FlowFile incomingFlowFile = processSession.get();
+
+            if (incomingFlowFile == null && context.hasNonLoopConnection()) {
+                return;
+            }
+            fetchSize = context.getProperty(IOTDB_QUERY_FETCH_SIZE).evaluateAttributeExpressions(incomingFlowFile).asInteger();
+            if ( (incomingFlowFile != null ? incomingFlowFile.getSize() : 0) == 0 ) {
+                if ( context.getProperty(IOTDB_QUERY).isSet() ) {
+                    query = context.getProperty(IOTDB_QUERY).evaluateAttributeExpressions(incomingFlowFile).getValue();
+                } else {
+                    String message = "FlowFile query is empty and no scheduled query is set";
+                    getLogger().error(message);
+                    incomingFlowFile = processSession.putAttribute(incomingFlowFile, "iotdb.error.message", message);
+                    processSession.transfer(incomingFlowFile, REL_FAILURE);
+                    return;
+                }
```

Comment:
This set of conditionals is deeply nested. Although formatting the query from FlowFile content is an interesting use case, the general ability to construct a query based on FlowFile attributes seems like it should cover common operations. Removing support for reading the query from the FlowFile would help simplify the logic. It also could be helpful to abstract some of the logic, where possible, into a separate `getQuery` method.